common/djangoapps/embargo/api.py

```diff
@@ -0,0 +1,123 @@
+"""
+The Python API layer of the country access settings. Essentially the middle tier of the project, responsible for all
+business logic that is not directly tied to the data itself.
+
+This API is exposed via the middleware(emabargo/middileware.py) layer but may be used directly in-process.
+
+"""
+import logging
+import pygeoip
+
+from django.core.cache import cache
+from django.conf import settings
+from embargo.models import CountryAccessRule, IPFilter, RestrictedCourse
+
+log = logging.getLogger(__name__)
+
+
+def _check_ip_lists(ip_addr):
+    """
+    Check whether the user is embargoed based on the IP address.
+
+    Args:
+        ip_addr (str): The IP address the request originated from.
+
+    Returns:
+        A True if the user is not blacklisted, otherwise False
+
+    """
+    # If blacklisted, immediately fail
+    if ip_addr in IPFilter.current().blacklist_ips:
+        return False
+
+    # If we're white-listed, then allow access
+    if ip_addr in IPFilter.current().whitelist_ips:
+        return True
+
+    return True
+    # If none of the other checks caught anything,
+    # implicitly return True to indicate that the user can access the course
+
+
+def get_user_country_from_profile(user):
+    """
+    Check whether the user is embargoed based on the country code in the user's profile.
+
+    Args:
+        user (User): The user attempting to access courseware.
+
+    Returns:
+        user country from profile.
+
+    """
+    cache_key = u'user.{user_id}.profile.country'.format(user_id=user.id)
+    profile_country = cache.get(cache_key)
+    if profile_country is None:
+        profile = getattr(user, 'profile', None)
+        if profile is not None and profile.country.code is not None:
+            profile_country = profile.country.code.upper()
+        else:
+            profile_country = ""
+        cache.set(cache_key, profile_country)
```

Comment:
There's a default cache expiration that's set fairly high (many minutes).  In this case, I think that's okay, since profile country can't change post-registration.  If T&L gives students the ability to change this in their profiles, we can add cache invalidation when the user makes the change.